### PR TITLE
[shelly] Support for Plus Smoke, Plus Plug-S/IT/UK/US, Plus Dimmer US, Pro 3EM;  fix Gen1 sensor initialization

### LINKED ISSUE #14346: [shelly] Add support for Shelly Pro 3EM
Support for the new Shelly Pro 3EM would be great.
I couldn't find a repo for the shelly binding where there would be some code I could work on and trigger a PR miself, so all I can do right now is open this issue and hope for someone to pinpoint me to where the dev is being done.

Since the two pages (http://<device ip>/settings and http://<device ip>/status) don't exist in Pro devices, I don't know where to start.

### LINKED ISSUE #14360: [shelly] CoIoT Temperature readings of 999 should be excluded
## Expected Behavior
Readings of 999 shoudln't be recorded into openhab:

> Temperature (in C) measured by an external sensor (e.g. Shelly 1 with DS1820/DHT22 add-on) or ambient temperature (e.g. Shelly H&T). This is a separate property since some devices also measure internal device temperature and it should be distinguished between the two. 999 if no valid temperature reading is available at the moment.

https://shelly-api-docs.shelly.cloud/gen1/docs/coiot/v1/CoIoT%20for%20Shelly%20devices%20(rev%201.0)%20.pdf

## Current Behavior
999 degrees are stored for shelly sensors using CoIoT

## Possible Solution
Check value isn't 999
https://github.com/rossbuggins/openhab-addons/pull/1/files
Is it possible to record N/A / Null / undef at this point instead? 

## Steps to Reproduce (for Bugs)
Shelly sensor, using CoIoT, with temperature (I'm seeing this on a Shelly 1)

## Context
Skews all data / graphs.

## Your Environment
3.4.1

### LINKED ISSUE #14539: [shelly] meter#current with wrong precision for EM, 3EM and all Plus/Pro relays
The channel meter#current uses 3 decimal digits. However, the binding rounded with 1 only so the value shows up with 3, but lost precision. Fix: New binding uses 3 decimal digits for rounding.

### LINKED ISSUE #14540: [shelly] Battery powered devices like Button don't connect after OH restart
After a OH restart the button doesn't connect again

This is fixed in PR #14532

### LINKED ISSUE #14551: [shelly] Fix NPE in Shelly Manager
Opening Shelly Manager could led into NPEs depending on active devices and types (esp. Plus/Pro). Some more checking has been added to prevent this.

### PR DESCRIPTION
**Pull request by markus7017**
This PR adds support for Shelly Plus Smok, Plus Plug-S, IT, UK, US, Pro 3EM
fixes CoAP initialization for Gen1 devices, ignores temp=999 on channel updates (invalid temp)

Fixes #14360
Fixes #14346
Fixes #14539
Fixes #14540
Fixes #14551

### CODE REVIEW

**Review by DM6JM — APPROVED** [review]
Did check through all the code and didn't find anything asking for change. Tested on my hab with newly added Pro3EM and didn't experience any problems.

**Comment by markus7017** [comment]
thanks for the feedback, @fwolter will do the review

**Comment by markus7017** [comment]
@fwolter when are you ready to start the review? The PR includes various enhancements for devices already on the market

**Comment by fwolter** [comment]
My priorities shifted a bit towards other project. Not sure if I'm gonna do the review.

**Comment by markus7017** [comment]
@openhab/add-ons-maintainers Is there anybody else who could do the review? The PR is small, so not a big task

**Review by DM6JM — APPROVED** [review]
Reviewed as requested, but I am not too deep into this binding. Works without issues on my system and the changes reviewed didn't unveil anything lokkung wrong.

**Comment by markus7017** [comment]
@openhab/add-ons-maintainers Could someone do this 500 line review? There are already the next devices to be integrated, but I want to keep the PR small ;-)

**Comment by markus7017** [comment]
@kaikreuzer Is the maintainer team on vacation. Could anybody inform what's going on?

File: bundles/org.openhab.binding.shelly/README.md [hunk 1]
```diff
@@ -80,23 +80,29 @@ The binding provides the same feature set across all devices as good as possible
...
-| shellypro1          | Shelly Pro 1 with 1x relay                               | SPSW-001XE16EU |
-| shellypro1pm        | Shelly Pro 1 PM with 1x relay + power meter              | SPSW-001PE16EU |
-| shellypro2-relay    | Shelly Pro 2 with 2x relay, relay mode                   | SPSW-002XE16EU |
-| shellypro2pm-relay  | Shelly Pro 2 PM with 2x relay + power meter, relay mode  | SPSW-002PE16EU |
-| shellypro2pm-roller | Shelly Pro 2 PM with 2x relay + power meter, roller mode | SPSW-002PE16EU |
+| shellypro1          | Shelly Pro 1 with 1x relay                               | SPSW-001XE16EU, SPSW-101XE16EU, SPSW-201XE16EU |
+| shellypro1pm        | Shelly Pro 1 PM with 1x relay + power meter              | SPSW-001PE16EU, SPSW-101PE16EU, SPSW-201PE16EU |
+| shellypro2-relay    | Shelly Pro 2 with 2x relay, relay mode                   | SPSW-002XE16EU, SPSW-102XE16EU, SPSW-202XE16EU |
+| shellypro2pm-relay  | Shelly Pro 2 PM with 2x relay + power meter, relay mode  | SPSW-002PE16EU, SPSW-102PE16EU, SPSW-202PE16EU |
+| shellypro2pm-roller | Shelly Pro 2 PM with 2x relay + power meter, roller mode | SPSW-002PE16EU, SPSW-102PE16EU, SPSW-202PE16EU |
```
Comment (jlaur) [inline bundles/org.openhab.binding.shelly/README.md:102]: Could alignment also be restored here?

File: bundles/org.openhab.binding.shelly/README.md [hunk 2]
```diff
@@ -374,7 +380,7 @@ A new alarm will be triggered on a new condition or every 5 minutes if the condi
 | TEMP_OVER  | Above "temperature over" threshold                                                               |
 | VIBRATION  | A vibration/tamper was detected (DW2 only)                                                       |
 
-Refer to section [Full Example:shelly.rules](#shellyrules) for examples how to catch alarm triggers in openHAB rules
+Refer to section Full Example for examples how to catch alarm triggers in openHAB rules
```
Comment (jlaur) [inline bundles/org.openhab.binding.shelly/README.md:383]:
```suggestion
Refer to section [Full Example](#full-example) for examples how to catch alarm triggers in openHAB rules.
```

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java [hunk 3]
```diff
@@ -292,7 +275,7 @@ public static boolean updateMeters(ShellyThingInterface thingHandler, ShellySett
                             updated |= thingHandler.updateChannel(groupName, CHANNEL_EMETER_VOLTAGE,
                                     toQuantityType(getDouble(emeter.voltage), DIGITS_VOLT, Units.VOLT));
                             updated |= thingHandler.updateChannel(groupName, CHANNEL_EMETER_CURRENT,
-                                    toQuantityType(getDouble(emeter.current), DIGITS_VOLT, Units.AMPERE));
+                                    toQuantityType(getDouble(emeter.current), DIGITS_KWH, Units.AMPERE));
```
Comment (jlaur) [inline bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java:278]: Please introduce a new `DIGITS_AMPERE` constant, or rename to `DIGITS_3` if you don't want the abstraction.

**Review by jlaur — CHANGES_REQUESTED** [review]
Thank you. Only some minor comments added.

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/shellyGen2_relay.xml [hunk 4]
```diff
@@ -219,6 +233,26 @@
 		<config-description-ref uri="thing-type:shelly:relay-gen2"/>
 	</thing-type>
 
+	<thing-type id="shellypro3em">
+		<label>Shelly Pro 3EM</label>
+		<description>@text/thing-type.shelly.shellypro3em.description</description>
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/shellyGen2_relay.xml:238]: The label is the same for EN+DE, whereas the description is language specific

File: bundles/org.openhab.binding.shelly/README.md [hunk 5]
```diff
@@ -80,23 +80,29 @@ The binding provides the same feature set across all devices as good as possible
 | -------------------- | -------------------------------------------------------- | ------------- |
 | shellyplus1          | Shelly Plus 1 with 1x relay                              | SNSW-001X16EU |
 | shellyplus1pm        | Shelly Plus 1PM with 1x relay + power meter              | SNSW-001P16EU |
-| shellyplus2pm-relay  | Shelly Plus 2PM with 2x relay + power meter, relay mode  | SNSW-002P16EU |
-| shellyplus2pm-roller | Shelly Plus 2PM with 2x relay + power meter, roller mode | SNSW-002P16EU |
+| shellyplus2pm-relay  | Shelly Plus 2PM with 2x relay + power meter, relay mode  | SNSW-002P16EU, SNSW-102P16EU |
+| shellyplus2pm-roller | Shelly Plus 2PM with 2x relay + power meter, roller mode | SNSW-002P16EU, SNSW-102P16EU |
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md:84]: fixed

File: bundles/org.openhab.binding.shelly/README.md (hunk 1 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md:102]: fixed

File: bundles/org.openhab.binding.shelly/README.md (hunk 2 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/README.md:383]: changed

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java (hunk 3 above)
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java:278]: new DIGITS_AMPERE

**Comment by markus7017** [comment]
@jlaur thanks for picking up this PR; changes are applied

File: bundles/org.openhab.binding.shelly/README.md (hunk 2 above)
Comment (jlaur) [reply bundles/org.openhab.binding.shelly/README.md]: Are you sure the link will work? I'm no expert in Markdown, so I don't know if the dot will be part of the heading id? Also, do you really want it to show as "Full Example:shelly.rules", it seems a little compressed?

Comment (jlaur) [reply bundles/org.openhab.binding.shelly/README.md]: Now noticed that you changed it back, so was easy to test on existing documentation page - the link is broken, but this link will work: https://www.openhab.org/addons/bindings/shelly/#shelly-rules

Comment (J-N-K) [reply bundles/org.openhab.binding.shelly/README.md]: @markus7017 Did you see this comment? It seems your changes are different. Can you comment on that?

**Comment by markus7017** [comment]
@jlaur anything missing?

**Comment by J-N-K** [comment]
Did you see my comment from yesterday? Otherwise I would have merged since all other issues seem to be addressed.

**Comment by jlaur** [comment]
> anything missing?

Only this: https://github.com/openhab/openhab-addons/pull/14532#discussion_r1193021075

**Comment by markus7017** [comment]
ah, ok, fixed

File: bundles/org.openhab.binding.shelly/README.md (hunk 2 above)
Comment (jlaur) [reply bundles/org.openhab.binding.shelly/README.md]: @markus7017 - now the link should work, but did you want a deep link to `#shelly-rules`? See suggestion above.

Comment (jlaur) [reply bundles/org.openhab.binding.shelly/README.md]: Leaving it as is to reach M3.

**Pull request merged**